### DISCUSSION TITLE
Dynamic linking non-zend extensions and how to statically link with a zend extension ?

### DISCUSSION THREAD

**Discussion by henrywood** [crazywhalecc/static-php-cli#369] [Q&A] [answered]
I just stumbled upon static-php-cli. Looks promising and cool.
My situation is that my company has a couple of PHP extensions (one of which is a zend_extension) and the two others are regular extensions.
I assume that you statically link all extensions to static-php-cli (since people ask for extensions being added).

Now for my questions:

1. Couldn't you just load (non-zend) extensions using dl() at runtime ?
2. Could you show an example of how to compile static-php-cli statically with my company's zend_extension

Thanks for your effort on static-php-cli !

**Comment by crazywhalecc** [comment, marked as answer]
> Couldn't you just load (non-zend) extensions using dl() at runtime ?

Actually it depends on your OS. Linux cannot load dynamic extension at runtime in static binary (linux kernel limitations). Windows and macOS do support loading dynamically with static binary (same to [ffi](https://static-php.dev/en/guide/extension-notes.html#ffi)).

But if you are going to run statically linked PHP binary on specific distro, like ubuntu, debian (based on glibc), it's possible to make a NOT-PURELY-STATIC linux executable while also supporting portability.

> Could you show an example of how to compile static-php-cli statically with my company's zend_extension

For normal extension, not hard, you can refer to #366 and [Docs-Developing](https://static-php.dev/en/develop/) section. In short, it needs to add `source`, add `lib` if it needs deps, add `ext` and some configure params. Sometimes you need to patch some code.

But I haven't added external Zend extensions yet. The only supported zend extension for SPC is Opcache. The Opcache support is done thanks to phpmicro's [patch](https://github.com/dixyes/phpmicro/blob/master/patches/static_opcache_83.patch). If you are implementing it this might help you

**Reply by henrywood** [reply]
Thanks - I'll have a look !
